Changes to sidebar for Application Layout

weiqingtoh opened the pull request:
Hi all, need to get some feedback, this involves code so I decided to do it as a PR instead. I wanted to do up the User badge on top of the sidebar; but when figuring out the mobile view, I thought it would be good to resolve this first. 

Within `app/themes/default/views/layouts` the site-wide navbar is implemented, as well as elements  for collapsed navbar and sidebar. So fundamentally users might need to deal with two menus, navbar and sidebar. Currently the mobile view for sidebar menus is not implemented accurately yet, which is the purpose of this discussion. 

*Current Image*
<img width="200" alt="screen shot 2015-11-09 at 3 44 59 pm" src="https://cloud.githubusercontent.com/assets/4353853/11087101/3ba3e6d8-8892-11e5-9aa6-9a1a7d1a4c6d.png">

If we adhere to the current helper structure, we can either: 
1. Get each sidebar element to handle their own collapse toggles, or
2. Place the collapse toggles (for navbar and sidebar) in the default layout.

<img width="350" alt="screen shot 2015-11-09 at 7 06 50 pm" src="https://cloud.githubusercontent.com/assets/4353853/11087594/9aa70c06-8896-11e5-92cd-16e0359b52ec.png">

IMO, we can go with option 1 for now to keep things simple. 

Ideally, the best is to nest the sidebar menu within the collapsed menu of the navbar (or vice versa). My guess is that this will require some severe refactoring. *See below pic:*

<img width="200" alt="screen shot 2015-11-09 at 11 17 43 pm" src="https://cloud.githubusercontent.com/assets/4353853/11087614/d1556a5e-8896-11e5-8e29-326ae2fb4633.png">


I’m not sure which is the best way forward, in terms of code and UI/UX. Anyone wants to chip in their views? 

---

In this PR, a few things are done (for everyone to see what I'm trying to mean): 
- Standardise the grid layout
- Implement option 1 for `app/views/layouts/_sidebar.html.slim` in `app/themes/default/views/layouts/default.html.slim`
- Implement option 2 for `app/views/layouts/course.html.slim`

*Not to merge PR yet, have to clean up stuff*

weiqingtoh commented:
@lowjoel @allenwq Any comments? Otherwise I will go with Option 1 first. Or is anyone doing themes, maybe can consult with them?

allenwq commented:
@weiqingtoh I'm fine with both, I think you can go option 1 first since it's more straightforward to implement.

weiqingtoh commented:
@allenwq please take a look, went with option 1 for now.

allenwq commented on app/views/layouts/_sidebar.html.slim:
```diff
@@ -1,3 +1,12 @@
+        span.sr-only
+          = t('layout.navbar.toggle_sidebar')
+        span.glyphicon.glyphicon-chevron-down aria-hidden='true'
+      = render partial: 'layouts/user', locals: { user: current_user }
```
Think we should display `course_user` here ?

weiqingtoh replied:
Am planning to implement `course_user` and the user badge in the next PR once I settle the sidebar stuff here.

allenwq replied:
Sure

lowjoel replied:
I don't understand why yield was removed?

lowjoel replied:
Why does every sidebar involve the user layout?

weiqingtoh replied:
Using this as the default

weiqingtoh replied:
That area is the navbar header. User layout seems to fit nicely for now cause it has a profile pic and a link to user. 

Will probably replace this when I implement the user badge. Course users should see a badge, and admins see something more basic.

lowjoel replied:
that doesn't make sense right -- you're using the sidebar in the instance admin and system admin and all other places where there's no notion of a course user.

allenwq commented on app/helpers/application_sidebar_helper.rb:
```diff
@@ -20,7 +20,7 @@ def sidebar!
   # @param [Array<String>] classes An array of classes to apply to the sidebar container.
   # @return [String] The buffer containing the markup for the sidebar.
-  def sidebar(classes: ['col-xs-7', 'col-sm-3', 'col-md-2'])
+  def sidebar(classes: ['col-lg-2', 'col-md-3', 'col-sm-4'])
```
Do we need `col-xs-` ?

weiqingtoh replied:
Was thinking of this, but when I tried viewing the page on `xs`, the sidebar takes up too much space. I'm proposing to collapse it into the navbar as seen in option 1 above, and because of that I didn't add the `col-xs-`

allenwq commented on spec/helpers/application_helper_spec.rb:
```diff
@@ -14,18 +14,27 @@
-        helper.sidebar do
-          ''
+      context 'when a user is signed in' do
+        let(:test_user) { User.new(name: 'user name') }
```
It's better to use FactoryGirl to build the user: `build(:user)`

allenwq commented on spec/helpers/application_helper_spec.rb:
```diff
@@ -14,18 +14,27 @@
-        result = helper.sidebar do
-          'Test Contents'
+        it 'shows the name of the user' do
+          expect(helper.sidebar { '' }).to include('user name')
```
Change to `test_user.name` ?

allenwq commented:
@lowjoel This PR looks good to me, do you have more comments ?

weiqingtoh commented:
@allenwq Updated! @lowjoel let me know if any comments too! 

I'm not sure how submodules work yet, but if I need to do a change to `coursemology-theme`, will do so after this is merged. 

Thanks!

lowjoel commented on app/views/layouts/course.html.slim:
```diff
@@ -2,12 +2,10 @@
-          = sidebar_items(controller.sidebar_items(type: :admin))
+    = sidebar do
+      = sidebar_items(controller.sidebar_items(type: :normal))
+      h5 = t('.administration')
```
I think semantically this is a bit strange also. the administration sidebar is actually a different set from the rest of the course menu, right?

weiqingtoh replied:
I see the sidebar method here more like a helper sidebar wrapper to help with styling. 

If I understand you correctly, this admin sidebar is the course admin sidebar. Not else how else we can organise this? Nesting in within the sidebar so that it all collapses when in mobile view.

lowjoel replied:
I see it as semantic -- HTML should be semantic, styling is a layer over that. Consider this case: what if there are multiple sidebars on a page?

lowjoel commented on spec/helpers/application_helper_spec.rb:
```diff
@@ -14,18 +14,27 @@
-        helper.sidebar do
-          ''
+      context 'when a user is signed in' do
+        let(:test_user) { build(:user) }
```
`build_stubbed`

lowjoel commented:
Try http://www.bootply.com/69863 -- does that look like it's different from us?

weiqingtoh commented:
@lowjoel that particular example doesn't handle cases when there is a mobile view though. You are thinking of adapting a similar HTML structure?

lowjoel commented:
How different is our current structure? I resized my viewport down, seems to collapse everything down?

weiqingtoh commented:
When you say your resize down and it seems to collapse, which particular page are you talking about?

In the current structure, _course-sidebar_ is implemented differently from _admin-sidebar_. 

For_course-sidebar_, it does collapse, but nothing implemented for mobile viewport yet. 
For _admin-sidebar_, the collapse trigger appears at the global navbar (see the first picture in the inital post). However, that button is not implemented correctly as: 1) The sidebar does not collapse, 2) The additional collapse button does not trigger anything. 

I'm trying to standardise the sidebar behaviour in this PR - currently setting it to collapse with mobile viewport, but if we do decide to make a smaller sidebar instead, it will be easier to do when we have a `sidebar` wrapper.

*P.s: Sorry, let me wrap my head around the html and semantic structure, will reply once I get a better understanding of that.*

lowjoel commented:
Oh, definitely, I think the consistency needs to be fixed. There's no need to fix the sidebars to all collapse nicely in this PR. I'm just suggesting how this problem might be dealt with.

My main concern now is that the sidebar has a user profile on every single occurrence.

weiqingtoh commented:
We could do a generic implementation (say just 'Menu', or nothing at all) and allow views to specify if a custom sidebar-header when calling the `sidebar` helper method.

Or am I going off tangent for this discussion?

lowjoel commented:
Why does a sidebar need a header? If it's for the user profile, that doesn't make the HTML semantic. If it's for the course admin sidebar, sure.

weiqingtoh commented:
Actually the sidebar header was to give a default view when the sidebar collapses with xs viewport (as we would need to add the button to trigger the drop down for the sidebar). I was planning to implement the user badge in the normal desktop view as well - Not sure whether that made sense. 

Anyway, can I check if this collapse was the intended behaviour of the sidebar, or did you have some other plans on how it should function?

lowjoel commented:
I'd imagine that eventually it'd look like option 3, the page loads, but the sidebar becomes part of the main nav menu.

weiqingtoh commented:
@lowjoel After the chat, I'm not sure if there is a better design, so I'm reverting back to the old code. The PR just fixes and standardises some of the sidebars for now. Let me know if you have some ideas that make more sense?

lowjoel commented on app/views/layouts/system_admin_instance.html.slim:
```diff
@@ -1,11 +1,13 @@
+      = sidebar do
+        ul.nav.nav-pills.nav-stacked
+          li
+            = link_to(t('system.admin.instance.navbar.users'), admin_instance_users_path)
```
@weiqingtoh should this be one link-to per li, or one li for every link?

lowjoel commented:
@weiqingtoh can you add a screenshot of the new implementation?

weiqingtoh commented:
**Desktop View**
<img width="500" alt="screen shot 2015-11-23 at 2 34 45 pm" src="https://cloud.githubusercontent.com/assets/4353853/11331056/5767ee84-91f0-11e5-9729-a8b4ee0c54ce.png">

**Mobile View** 
_(Left is default mobile view, middle is when expanding collapsed course sidebar, right is when expanding collapsed admin navbar)_
<img width="500" alt="combined" src="https://cloud.githubusercontent.com/assets/4353853/11331042/27058ae4-91f0-11e5-8c60-14b4f524fad8.png">

Nothing new actually, just added navbar-default to the sidebar to have some grey background, and fixed the collapse buttons.

lowjoel commented on app/views/layouts/course.html.slim:
```diff
@@ -3,8 +3,8 @@
-      nav role='navigation'
-        div.collapse.navbar-collapse#course-navigation-navbar
+      nav.navbar-default.sidebar role='navigation'
+        div.collapse.navbar-collapse#site-navigation-sidebar
```
last question, why is this site navigation sidebar when it's the course template?

lowjoel commented on app/views/layouts/user_admin.html.slim:
```diff
@@ -1,12 +1,14 @@
+      = sidebar do
+        ul.nav.nav-pills.nav-stacked
+          li
+            = link_to(t('user.admin.navbar.profile'), edit_user_profile_path)
```
like this is inconsistent vs the system admin instance sidebars

weiqingtoh commented:
Apologies, standardised the `li`, and decided to change the id to `#sidebar` instead. I think that will suffice for now.

weiqingtoh commented:
Okay I hope this is okay. Will need to make a PR to the themes page to fix the default layout as well.

lowjoel commented on app/views/layouts/_sidebar.html.slim:
```diff
@@ -1,3 +1,4 @@
-    = yield
+div class=sidebar_classes
+  nav.navbar-default.sidebar role='navigation'
+    div.collapse.navbar-collapse#site-navigation-sidebar
```
This can't have an ID because e.g. in Course there are two of these, because sidebar_items are called twice.

weiqingtoh replied:
This partial is not involved with `sidebar_items`. The current method for multiple sidebars is for the layout to handle it itself.

Only the `sidebar` helper method use this, which for now, is called at most once. Should I add that into the comments or?

lowjoel replied:
Okay, I'm a bit rusty on this part, got it.

sidebar creates a new sidebar. sidebar_items actually creates the items inside the sidebar...

lowjoel replied:
If you can explain it better in your next set of changes that would be awesome
